### ISSUE TITLE
Decide hierarchy of analysis classes/methods

### ISSUE THREAD

**Issue by hmacdope** [MDAnalysis/solvation-analysis#28]
Some of our proposed functionality follows on directly from other functionality and some does not.

This means we need to discuss how to **compartmentalise** functionality into classes and methods many of which will subclass analysis base. 

I thought I would make this issue as an open discussion area for how we want to compartmentalise /hierarchicalize our functionality.  I think the best way to proceed is if @orioncohen can lay out how he sees it and then @richardjgowers @IAlibay and any @MDAnalysis/gsoc-mentors can weigh in. We may also need to meet to discuss this. :)

**Comment by orioncohen** [comment]
I'll write up a suggested layout and post it here! I will also use the plan as a basis for a second blog post.

**Comment by hmacdope** [comment]
Sounds good!

**Comment by orioncohen** [comment]
The Solute class will be will be instantiated by specifying a solute AtomGroup and several solvent AtomGroups. Solvents will be specified with a dictionary where the keys are the solvent identifiers (strings) and the values are the AtomGroups of the solvents. By forcing users to tag all of their solvent types, it will be easier to provide sensible output. Instantiating a solute will look something like this:

```javascript
li_ion_group = u.select_atoms('name Li')
solvent1 = u.select_atoms('name Solvent1')
solvent2 = u.select_atoms('name Solvent2')
li_solute = Solute(li_ion_group, {name1: solvent1, name2: solvent2...})
```

Determining the solvation radius requires computing several RDFs, which can be time consuming. Rather than calculate the RDF's when the Solute is instatiated, which would be slow, I will let the user compute the RDF's later on. Once they are computed, the user can view the RDF's and the identified solvation shell to validate that they are reasonable.

```python
li_solute.prepare()
li_solute.rdf_plots[0]  # return rdf plot to validate solvation radii
```

Optionally, if the solvation radii are known, the user can specify them rather than having them computed automatically. If only some radii are known, they can specify only some of the radii by including False elements in the list. The relevant RDFs will still be calculated but now the specified solvation radius will be used instead of the calculated one.

```python
li_solute.prepare([3.1, 4.2])
li_solute.prepare([3.1, False])
li_solute.rdfs[0]
```

Finally, if users would like to use a different kernel for identifying the first solvation shell, they can set it at instantiation or later on.

```python
li_solute = Solute(li_ion_group, {name1: solvent1, name2: solvent2...}, radius_kernel=some_kernel_function)
```
OR
```python
li_solute = Solute(li_ion_group, {name1: solvent1, name2: solvent2...}
)
li_solute.kernel = some_kernel_function
```
THEN
```python
li_solute.prepare()
```

Once the solvation radii are calculated, the solvation shells for each ion can be calculated. All other analysis depend on the solvation shells, so computing them is the central operation of this package. It is also the most expensive part, so it makes sense to calculate once and then pass the data to other analysis. This necessitates the existence of some core data structure that represents solvation. It should be complete, minimal, and fast because all other analyses will rely on it. Once the li_solute is prepared, the user will call `li_solute.run()` to calculate the core data structure. Once this is done, all of the analysis are cheap and they will be automatically instantiated as attributes of the Solute class.

```python
li_solute = Solute(li_ion_group, {name1: solvent1, name2: solvent2...})
li_solute.prepare()
li_solute.run()
li_solute.some_analysis_class
```

However, having every analysis rely on the existence of Solute is not very modular, so instead, I will also allow the analysis classes to be created outside Solute. They will have two constructors that allow them to be instantiated from the solute and solvating species or the core solvation data structure. This will make them modular and provide a convenient shared interface for all analysis classes.

```python
coordination = CoordinationNumbers(solute, {name1: solvent1, name2: solvent2...}, [3.44, 3.11, ...])
coordination.run()
speciation = IonSpeciation(solute, {name1: solvent1, name2: solvent2...}, [3.44, 3.11, ...])
speciation.run()
pairing = IonPairing(solute, {name1: solvent1, name2: solvent2...}, [3.44, 3.11, ...])
pairing.run()
```
or alternatively, they can be instantiated from the precalculated solvation data, which is what would happen inside the Solute class.
```python
solvation_data = li_solute.core_solvation_data_struct
coordination = CoordinationNumbers.from_prepared(solvation_data)
speciation = IonSpeciation.from_prepared(solvation_data)
pairing = IonPairing.from_prepared(solvation_data)
```
This will also give a shared interface for all of the solvation analysis objects, they can all be implemented by specifying the solute, solvents, and for some, the solvation radii.

I am not quite sure what the core data structure representing solvation should be. It will be computed for every ion across many frames so it should be relatively lightweight. It should also include enough information for other analysis classes to use it without re-computing anything expensive.

For a single ion it could be a  list of dicts, where each solvating molecule has a dict that specifies the distance of the closest atom, the atoms in the group, and the tag that the user specified for that solvating molecule.
```python
[{dist: distance1, atoms: group1, tag: tag1}, 
{dist: distance2, atoms: group2, tag: tag2}, ...]
```

I'd love feedback on this plan. Particularly on what the core data structure could be and how to efficiently calculate it. Please let me know if you have any questions!

**Comment by orbeckst** [comment]
Two quick comments: 
1. Is there a way in your scheme to save intermediate/expensive results and then use these to re-instantiate your analysis classes? In particular not using Python pickles but "real" data formats such as CSV, HDF5, ...?
2. Generally, "lists of dicts" (arrays of data structures) are lower performance than "dicts of lists" (multiple arrays with the long axis along the trajectory, i.e. each array is N_frames long and you create as many arrays as necessary). Alternatively, you can also typically do a "long dataframe" (also know as "[tidy data](https://tidyr.tidyverse.org/articles/tidy-data.html)") where each row corresponds to a single observation and all columns except one contain "keys". This format requires extra storage but is easy to process (especially with indexing and group-by type operations) and is very flexible. In numpy you can use arrays with custom dtypes (or rec arrays although they can be a bit slow).

**Comment by orbeckst** [comment]
I like the way you instantiate `Solute` as this is pretty clean and expressive.

I would continue using the solvent names. E.g. when specifying known radii, I would not rely on an existing order (which is not guaranteed by dicts) but use kwargs
```python
li_solute.prepare(name1=3.1)
```
Alternatively, provide a known radius when you set up the `Solute`:
```python
solute = Solute(ag_solute, {name1: {"atoms": ag1, "radius": 3.1}, {name2: {"atoms": ag2}})
```
or
```python
solute = Solute(ag_solute, {name1: ag1, name2: ag2}, radii={name1: 3.1})
```

Given that `.prepare()` is an expensive function, I'd use a verb that makes this more clear. I am not quite sure what, `run_prepare()` or `run_initial()` or `run_rdf` or `run_first` ... but `prepare()` sounds to me as if it should not take much time.

**Comment by lilyminium** [comment]
Just a suggestion: if the tidy dataframe sounds good to you but you want to avoid pandas, you could consider using xarray instead.

**Comment by hmacdope** [comment]
Hi Orion I like the way the plan is rolling out so far! Couple of comments to add to what @orbeckst and @lilyminium have said (thanks all!). 

* As Oliver said I would stick to accessing things using the names you have provided. so something like `li_solute.rdf_plots[0]` can become `li_solute.rdf_plots["solventname"]` (ie become a dictionary). I do really like the use of names as it encourages you to keep track of which species you are talking about.  
* radius kernel as a kwarg is a good idea!
* I think having three ways of doing the same thing will be a bit confusing for users. Ie having being able to calculate a coordination number by `li_solvation.CoordinationNumber(blah)` and `CoordinationNumber(Solute)` and `coordination = CoordinationNumbers.from_prepared(solvation_data)` is  somewhat redundant and will confuse users , as it can be hard to keep track of what is cross compatible with what etc etc. 
* On this note I think simplicity is best, if the user has 10 ways of doing something, they will often choose the least efficient way. Better to force people down the right path through design.
* I think you can introduce the modularity you want through library functions, ie have a function `_coordination_number()` in a file called something like `solvation_library.py` (much like you have already done for your `get_closest_n_mol` etc etc) and then all the different ways to use coordination number can call that under the hood. That way every data structure doesn't need to be able to talk to every other one and can just call library functions. 
* WRT the core data structure, its basically just set of tensor of distance matricies ie (N_rdfs *N_solute_atoms *N_solvent_atoms) There may be a way to store these after they are calculated by `InterRDF` if you can dive into the interior of the class. AFAIK after that is done almost everything is a lookup operation on that tensor (ie calculating a coordination number is selecting a subset of these and integrating etc etc). Happy to discuss here and in our upcoming meeting.

**Comment by orioncohen** [comment]
Thank you for the feedback everyone!!

To @orbeckst and @lilyminium suggestions:

- Yes it would not be difficult to save an intermediate representation that the Solute class could be instantiated from. I can probably use the core data structure itself to reinstantiate solute. 
- Thanks for the suggestion of a "tidy" dataframe, I think that would be cleaner than using a list / dictionary based representation. I'll check out xarray as @lilyminium suggested.
- I really like the idea of using solvent names going forward, that's much better than relying on indexing.

To @hmacdope's suggestions:

- Your point on multiple ways of doing things is well taken. I'll put the coordination, speciation, etc. analysis functions in a separate file and filter everything through Solute. I was trying to shoehorn the cheaper analyses into the AnalysisBase class but I don't think it's really appropriate. 
- I'm interested in hearing more about the idea of extracting everything from the RDF data as that would be a really cheap way of calculating the solvation shells. I do wonder if it would be preferable to have something more processed as the core data structure. For example, perhaps the non-solvating molecules could be filtered out and the remaining atoms could be organized by type and distance. Then the analysis operations would be easier and cheaper.

**Comment by hmacdope** [comment]
If we do go down that route I would try and keep the tensor indicies simple, or at least corresponding to the order in the `atomgroups` that way you don't have to undo / keep track of any ordering or masking ops to figure out the atom number. Regardless, lets get an example skeleton up and finesse from there :)

**Comment by orioncohen** [comment]
In the spirit of iterating quickly, I've started out with a familiar pandas array. For now, each observation is an atom in a solvation shell and the columns are "atom id", "solvated atom", "distance", "residue name", and "residue id". I believe this data structure is "tidy" which would make it easy to further process and analyze. I'll likely order the internal data structure by solvated atom to make it more readable. 

I am quite interested in xarray because I like the data abstraction but I don't want to get slowed down learning it. I'll revisit once I get some basic functionality out.

I am calculating the solvation shells using distances.capped_distances and numpy array operations, which seems to be quite fast so far. I am iterating through 10 frames of my test dataset (60 li ions, ~7000 atoms total) in less than 1 ms.

I am going to crosspost this on PR #29 with the new additions!

**Comment by hmacdope** [comment]
Awesome work. Long term however I am not a pandas superfan, however good to get things kicked off.
